README.md
# string_to_hex [![pub package](https://img.shields.io/badge/pub-0.2.2_null_safety-blue)](https://pub.dev/packages/string_to_hex)

A powerful conversion of [String] or/and Hash to HEX.
It returns a unique HEX, or a unique int of [Color()] per provided String/hash.
It's provided two methods [toHexString] and [toColor], which return a Hex-String, or integer of Color() respectively.

## Default:

This package return hex-string or hex-int-color, check bellow:

#### toColor()

Return a [int] of 'bit hex':
Its return is proper to use as hex [color int] in a Color()
For example: `... color: Color(StringToHex.toColor('a nice String')) ...`
then it'll generate and fill a hex-color int in it.

    return a hex-color.
    i.e: 0xFF353535
    or: 8787451701

##### how to:

you can use as:

```
  var myNiceColor = StringToHex.toColor('your nice string');
    ...
    color: Color(myNiceColor);
    ...
```

or:

```
...
color: Color(StringToHex.toColor('your nice string'));
...
```

#### toHexString()

Return a [String] of 'bit hex':
Its return is proper to your custom manipulating (as you lik better 🧑‍💻).

    Returns a String HEX with prefix '0xFF'
    i.e: '0xFF343434'

## Example:

[This example app shows all implementations](https://github.com/allansrc/string_to_hex/tree/master/example)

| App shot                                                                                                     |
| ------------------------------------------------------------------------------------------------------------ |
| <img src="https://github.com/allansrc/string_to_hex/blob/master/image/ezgif-4-ad8ddb7cb25a.gif" width="250"> |

![App Shots](https://github.com/allansrc/string_to_hex/blob/master/image/ezgif-4-ad8ddb7cb25a.gif)

Have a Nice Code!


CHANGELOG.md
## [0.0.1] - 2020-06-25.

- Initial release.

## [0.1.0] - 2020-06-27.

- Release upgrade.

## [0.1.1] - 2020-06-27.

- Update docs.

## [0.1.1+1] - 2020-06-27.

- fix image docs.

## [0.1.1+2] - 2020-06-27.

- fix image docs.

## [0.2.0] - 2020-07-05.

- Methods improv & docs.

## [0.2.1] - 2020-07-30.

- fix range error.

## [0.2.1+30] - 2020-07-30.

- Method call as `StringToHex.toColor()` replacing previewer call `StringToHex().toColor()`

## [0.2.2] - 2021-06-23.

- Migrate to null-safety


test/string_to_hex_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:string_to_hex/string_to_hex.dart';

void main() {
  test('adds one to input values', () {
    final calculator = Calculator();
    expect(calculator.addOne(2), 3);
    expect(calculator.addOne(-7), -6);
    expect(calculator.addOne(0), 1);
    expect(() => calculator.addOne(null), throwsNoSuchMethodError);
  });
}


example/lib/main.dart
import 'dart:math';

import 'package:flutter/material.dart';
import 'package:string_to_hex/string_to_hex.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
        visualDensity: VisualDensity.adaptivePlatformDensity,
      ),
      home: MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key, this.title}) : super(key: key);

  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  int _counter = 5;
  String _string = 'your nice string';
  String hexColor = 'The Hex String comes here';
  Color generatedColor;
  int generatedColorInt;

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        backgroundColor: Color(4278272051),
        title: Text(widget.title),
      ),
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            MaterialButton(
              child: Text('press to generate Color & HexString'),
              color: Color(generatedColorInt ?? 0xfff5f535),
              onPressed: () {
                setState(() {
                  hexColor = getRandomString(_counter);
                  _string = StringToHex().toHexString(_string);
                  generatedColor = Color(StringToHex().toColor(_string));
                  generatedColorInt = StringToHex().toColor(_string);
                });
              },
            ),
            Text(
              '$_string',
              style: Theme.of(context).textTheme.headline4,
            ),
            Text(
              '$hexColor',
              style: Theme.of(context).textTheme.headline4,
            ),
          ],
        ),
      ),
    );
  }
}

const _chars = 'AaBbCcDdEeFfGgHhIiJjKkLlMmNnOoPpQqRrSsTtUuVvWwXxYyZz1234567890';
Random _rnd = Random();

String getRandomString(int length) {
  return String.fromCharCodes(Iterable.generate(length, (_) => _chars.codeUnitAt(_rnd.nextInt(_chars.length))));
}


lib/string_to_hex.dart
library string_to_hex;

/// A powerful conversion of [String] or/and Hash to HEX.
/// It returns a unique HEX, or a unique int of [Color()] per provided String/hash.
/// It's provided two methods [toHexString] and [toColor], which return a Hex-String, or respectively.
class StringToHex {
  /// This class get RGB Color
  /// by doing bits shifts from color [0xFF0000], [0x00FF00], [0x0000FF]

  static int _getInt(str) {
    var hash = 5381;

    for (var i = 0; i < str.length; i++) {
      hash = ((hash << 4) + hash) + str.codeUnitAt(i) as int;
    }

    return hash;
  }

  /// Return a [String] of 'bit hex':
  /// Its return is proper to your custom manipulating (as you lik better 🧑‍💻).

  /// Returns a String HEX with prefix '0xFF'
  /// i.e: '0xFF343434'
  static String toHexString(str) {
    try {
      var hash = _getInt(str);
      var r = (hash & 0xFF0000) >> 16;
      var g = (hash & 0x00FF00) >> 8;
      var b = hash & 0x0000FF;

      var rr = r.toString();
      var gg = g.toString();
      var bb = b.toString();

      return '0xFF' +
          rr.substring(rr.length - 2) +
          gg.substring(gg.length - 2) +
          bb.substring(bb.length - 2);
    } catch (err) {
      print('Error: String Must be greater than range 2\n'
          '=========== hash string to hex ===========\n'
          '            string length = ${str.length}');
      return err.toString();
    }
  }

  /// Return a [int] of 'bit hex':
  /// Its return is proper to use as hex [color int] in a Color()
  /// For example: ``` ... color: Color(StringToHex().toColor('a nice String')) ... ```
  /// then it'll generate and fill a hex-color int in it.

  /// return a hex-color.
  /// i.e: 0xFF353535
  /// or: 8787451701
  static int toColor(str) {
    try {
      var hash = _getInt(str);
      var r = (hash & 0xFF0000) >> 16;
      var g = (hash & 0x00FF00) >> 8;
      var b = hash & 0x0000FF;

      var rr = r.toString();
      var gg = g.toString();
      var bb = b.toString();

      return int.parse('0xFF' +
          rr.substring(rr.length - 2) +
          gg.substring(gg.length - 2) +
          bb.substring(bb.length - 2));
    } catch (err) {
      print('Error: String Must be greater than range 2\n'
          '=========== hash string to hex ===========\n'
          '            string length = ${str.length}');
      rethrow;
    }
  }
}
